feat: improve tree-shaking by propagate const parameter

liuly0322 opened the pull request:
<!--
  ⚡️ katchow! We ❤️ Pull Requests!

  If you remove or skip this template, you'll make the 🐼 sad and the mighty god
  of Github will appear and pile-drive the close button from a great height
  while making animal noises.

  Pull Request Requirements:
  * Please include tests to illustrate the problem this PR resolves.
  * Please lint your changes by running `npm run lint` before creating a PR.
  * Please update the documentation in `/docs` where necessary

  Please place an x (no spaces - [x]) in all [ ] that apply.
-->

This PR contains:

- [ ] bugfix
- [x] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

List any relevant issue numbers:

<!--
If this PR resolves any issues, list them as

-  resolves #1234

where 1234 is the issue number. This will help us with house-keeping as Github will automatically add a note to those issues stating that a potential fix exists. Once the PR is merged, Github will automatically close those issues.

Starting each line with a dash "-" will cause GitHub to display the issue title inline.

If an issue is only solved partially or is relevant in some other way, just list the number without "resolves".
-->

### Description

<!--
  Please be thorough and clearly explain the problem being solved.
  * If this PR adds a feature, look for previous discussion on the feature by searching the issues first.
  * Is this PR related to an issue?
-->

Hi, all! This Pull Request trys to collect all function calls of top-level functions, so that we can determine if a param of the function is a literal to improve tree shaking (or be used to remove unused default params, which is not included in this PR)

The idea comes from #5435 , and related PRs: #4498 , #4510 , which are finally reverted by #4520 

This PR has changed the original code framework to some extent (to use a new approach to collect function use-def information), so I want to know whether the method of collecting all references in the PR and the placement position of the optimization pass are reasonable, thanks!

before:

```js
function fun1(enable) {
	if (enable) {
		return 'fun1';
	}
}

function fun2(enable) {
	if (enable) {
		return fun1(enable);
	}
}

console.log(fun2(true));
```

after:

```js
function fun1(enable) {
	{
		return 'fun1';
	}
}

function fun2(enable) {
	{
		return fun1();
	}
}

console.log(fun2());
```

liuly0322 commented:
Update: I think it will be pretty useful for library usage, I have just update an optimization:

before:

```js
function fun1(options) {
	if (options.enable) {
		return 'fun1';
	} else {
		console.log(222);
	}
}

function fun2(options) {
	if (options.enable) {
		return fun1(options);
	} else {
		console.log(111);
	}
}

console.log(
	fun2({
		enable: true
	})
);
```

output:

```js
function fun1(options) {
	if (options.enable) {
		return 'fun1';
	}
}

function fun2(options) {
	if (options.enable) {
		return fun1(options);
	}
}

console.log(
	fun2({
		enable: true
	})
);
```

and providing an `options` parameter is a common pattern for many librarys.

lukastaegert commented:
Thank you for your work! This is a really interesting approach. I will give you a more thorough review in the next days as I find time, maybe we can develop this in a slightly different direction.

lukastaegert reviewed (commented):
So, as promised, here is my review. This PR does indeed show quite some ingenuity. It is not [free of bugs](https://rollup-lvkk37y81-rollup-js.vercel.app/repl/?pr=5443&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGZ1bmN0aW9uJTIwZm9vKHgpJTIwJTdCJTVDbiUyMCUyMCUyRiUyRiUyMFRoZSUyMGV4cG9ydGVkJTIwZnVuY3Rpb24lMjBtaWdodCUyMGFsc28lMjBiZSUyMGNhbGxlZCUyMHdpdGglMjAlNUMlMjJmYWxzZSU1QyUyMiU1Q24lMjAlMjBpZiUyMCh4KSUyMGNvbnNvbGUubG9nKCd0cnVlJyklM0IlNUNuJTIwJTIwZWxzZSUyMGNvbnNvbGUubG9nKCdmYWxzZScpJTVDbiU3RCU1Q24lNUNuZm9vKHRydWUpJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlMkMlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJvdXRwdXQlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiU3RCU3RCU3RA==) but that would be fixable. Performance is also ok as your additional pass only scans top-level variables.

What I do not like, though, is that it does not tie well into the existing tree-shaking logic but rather appears to build its separate logic next to it. That also means that it
- needs to make a distinction to work only for top-level variables (which avoids performance issues)
- does not take side effects into account, e.g. [here](https://rollup-lvkk37y81-rollup-js.vercel.app/repl/?pr=5443&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmZ1bmN0aW9uJTIwZm9vKHgpJTIwJTdCJTVDbiUyMCUyMGlmJTIwKHgueSklMjBjb25zb2xlLmxvZygndHJ1ZScpJTNCJTVDbiUyMCUyMGVsc2UlMjBjb25zb2xlLmxvZygnZmFsc2UnKSU1Q24lN0QlNUNuJTVDbmZvbyglN0J5JTNBJTIwdHJ1ZSU3RCklM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSUyQyUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMm91dHB1dCUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTdEJTdEJTdE), it could also remove the `if(x.y)`, but it is retained because the property access `x.y` might have a side effect—which it obviously does not have.
- I am not sure how easy it is to build other features on top of this.

Also, there already IS a mechanism that connects call arguments and functions. Actually, two mechanisms. One is [`includeCallArguments`](https://github.com/rollup/rollup/blob/23d5282e016af67e78f29820037a3fb9c4526785/src/ast/scopes/ParameterScope.ts#L52-L93). This one is responsible for removing unused arguments from function calls like [here](https://rollupjs.org/repl/?version=4.13.0&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmZ1bmN0aW9uJTIwZm9vKHgpJTIwJTdCJTVDbiUyMCUyMGNvbnNvbGUubG9nKCdXaG8lMjBuZWVkcyUyMHgnKSU1Q24lN0QlNUNuJTVDbmZvbygndW51c2VkJyklM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSUyQyUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMm91dHB1dCUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTdEJTJDJTIydHJlZXNoYWtlJTIyJTNBdHJ1ZSU3RCU3RA==).
this might be one candidate where one could put additional logic for parameter tracking, i.e. the ParameterScope could tell the ParameterVariables which values were included. Note though that this might be a little late as until the point where the call is included, the function was evaluated without the parameter value, which means it can probably only assume the value to be "unknown", which might provide suboptimal results.

The other mechanism is [deoptimizeArgumentsOnInteractionAtPath](https://github.com/rollup/rollup/blob/23d5282e016af67e78f29820037a3fb9c4526785/src/ast/nodes/shared/FunctionBase.ts#L68-L86). This one is triggered very early, basically the first time a CallExpression is scanned for effects (which roughly corresponds to the first time Rollup determines this CallExpression to be in an included code path) or the first time it is included (for unconditionally included code), see [CallExpression.applyDeoptimizations](https://github.com/rollup/rollup/blob/23d5282e016af67e78f29820037a3fb9c4526785/src/ast/nodes/CallExpression.ts#L130-L138). This mechanism tracks mutations that happens in functions to mark call arguments as "deoptimized" if necessary, so that e.g. they [do not report wrong literal values](https://rollupjs.org/repl/?version=4.13.0&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmZ1bmN0aW9uJTIwZm9vKHgpJTIwJTdCJTVDbiUyMCUyMCUyRiUyRiUyMHRyeSUyMGNvbW1lbnRpbmclMjBvdXQlMjB0aGlzJTIwbGluZSU1Q24lMjAlMjB4LnklMjAlM0QlMjB0cnVlJTNCJTVDbiU3RCU1Q24lNUNuY29uc3QlMjBvYmolMjAlM0QlMjAlN0J5JTNBJTIwZmFsc2UlN0QlM0IlNUNuZm9vKG9iaiklM0IlNUNuJTVDbmlmJTIwKG9iai55KSUyMGNvbnNvbGUubG9nKCdPSycpJTNCJTVDbmVsc2UlMjBjb25zb2xlLmxvZygnRkFJTCcpJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlMkMlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJvdXRwdXQlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiU3RCU3RCU3RA==). You could use `deoptimizeArgumentsOnInteractionAtPath` to report possible values to the parameters.

You could then not limit the values to literal values alone, but if you do that, I would make sure not to track long lists of call parameters because then you might run into performance problems. Or again limit yourself to literal values for now. Rollup usually has the logic that if a function is reassigned, the value is assumed to be "unknown" because otherwise, we would often quickly run into exponential performance degradation issues.

There is another case to consider, though. What happens if a function is exported or assigned to a global variable? In that case, we must not make assumption about passed in parameters. Luckily, this case [can be easily detected as well](https://github.com/rollup/rollup/blob/23d5282e016af67e78f29820037a3fb9c4526785/src/ast/nodes/shared/FunctionBase.ts#L99-L106).

Would you consider re-implementing the feature in such a direction? One immediate benefit would be that it would not longer be limited to top-level functions and also work e.g. in IIFEs etc.

liuly0322 commented:
Thank you very much for your detailed explanation! I didn't know much about the internal details of rollup before, you help me a lot.

I have tried to refactor my code. `deoptimizeArgumentsOnInteractionAtPath` sounds great to collect function call information, but when taking re-export into consideration, I decide to collect all call information in `bind` step too (since `originalVariable` can be found for export default expression). The evaluation step is done in `include`.

I have updated the logic to fit in the tree-shaking framework. Now the updated version works for old tests, and a new test for don't remove function exports at top level. 

Below is a new problem:

When `hasEffects` is called, `if expression` and other expression will cache the condition result. later when I changed the parameter's calculated value in `include`, the if expression result is still cached, so it won't update. currently I remove the if expression cache, but I think it may not be the best solution. (there are other expressions with cache too)

Also, I am not sure how to remove `if(x.y)` with no side-effect in you [post](https://rollup-lvkk37y81-rollup-js.vercel.app/repl/?pr=5443&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmZ1bmN0aW9uJTIwZm9vKHgpJTIwJTdCJTVDbiUyMCUyMGlmJTIwKHgueSklMjBjb25zb2xlLmxvZygndHJ1ZScpJTNCJTVDbiUyMCUyMGVsc2UlMjBjb25zb2xlLmxvZygnZmFsc2UnKSU1Q24lN0QlNUNuJTVDbmZvbyglN0J5JTNBJTIwdHJ1ZSU3RCklM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSUyQyUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMm91dHB1dCUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTdEJTdEJTdE)

Thank you again!

lukastaegert commented:
> When hasEffects is called, if expression and other expression will cache the condition result. later when I changed the parameter's calculated value in include, the if expression result is still cached, so it won't update

That is why `IfStatement` is a `DeoptimizableEntity`, which is an entity that implements `deoptimizeCache`, and why `IfStatement` is passing itself as the [last parameter to `getLiteralValueAtPath`](https://github.com/rollup/rollup/blob/23d5282e016af67e78f29820037a3fb9c4526785/src/ast/nodes/IfStatement.ts#L127-L127) when getting the test value.

Caching literal values happens in many places to avoid repeated complicated lookups. That is why every Node that responds to `getLiteralValueAtPath` needs to handle the case when its value changes. Here is an example from `ConditionalExpression`:

Conditional expressions are ternaries `a ? b : c`. Initially, they are caching the literal value of `a`, or rather the implied resolution `b` or `c`, which they cache as the [`usedBrach`](https://github.com/rollup/rollup/blob/23d5282e016af67e78f29820037a3fb9c4526785/src/ast/nodes/ConditionalExpression.ts#L37-L37).

Now as conditional expressions are also expressions, it is possible to get their literal value, e.g. assume we have something (completely arbitrary and nonsensical) like

```js
if (a ? true : false) {
  ...
}
```

In this case if the value of `a` changes, not only do we need to remove the cached `unusedBranch` in the conditional, but also all caches in the if statement. The if statement passed itself as last parameter to `getLiteralValueAtPath`, which is called the `origin` parameter.
[In its own handler](https://github.com/rollup/rollup/blob/23d5282e016af67e78f29820037a3fb9c4526785/src/ast/nodes/ConditionalExpression.ts#L71-L80), ConditionalExpression first checks if it has a cached `usedPath`. Only in case it has one (which means there is a risk it might be deoptimized later), it will push the origin into its `expressionsToBeDeoptimized` array (if you want to see other examples, search for those arrays in other places).

Just like the if statement, `ConditionalExpression` [passes itself](https://github.com/rollup/rollup/blob/23d5282e016af67e78f29820037a3fb9c4526785/src/ast/nodes/ConditionalExpression.ts#L214-L214) to `getLiteralValueAtPath` when determining the value of `a`. When that value changes, the changing node will call [`deoptimizeCache`](https://github.com/rollup/rollup/blob/23d5282e016af67e78f29820037a3fb9c4526785/src/ast/nodes/ConditionalExpression.ts#L48-L59) on the ConditionalExpression, which the node will use to
- clear the cached value
- fully deoptimize the other branch (Which means here: Assume that some arbitrary mutations might have happened to the initial branch as we did not track those)
- deoptimize the cache of all `expressionsToBeDeoptimized`

Now I assume this is also what you need to do: Track which entities request the literal value and deoptimize them when your value changes.

Similar to how you implemented `getLiteralValueAtPath`, you need to implement `hasEffectsOnInteractionAtPath`, which tracks side effects from property accesses, calls and assignments. To avoid performance issues, this would only make sense if a function is really only called once. In that case, you would forward `hasEffectsOnInteractionAtPath` to the argument passed to the single call. If there is a second call, I would then just default to `true`. On the plus side, there are no caching problems here as `hasEffectsOnInteractionAtPath` is called again for every subsequent run if the return value is `false`.

liuly0322 commented:
Wow, thanks, now I learned how this great cache design works. https://github.com/rollup/rollup/pull/5443/commits/ac4794dde37e6fdad4a4772757db9dd99ad291f9 I have updated an implementation to use `deoptimizeCache`. There are few changes I think should be noticed:

1. since lattice may change twice, from top to some value and finally bottom, I remove the readonly `EMPTY_ARRAY` usage and set `expressionsToBeDeoptimized` to `[]` instead. And because lattice may change at most twice, this should be ok.
2. For `conditionalExpression` and `logicalExpression`, the old implementation will not trigger a re-computation and will just mark it as unknown. I update `isBranchResolutionAnalysed` to `false` in `deoptimizeCache` to fix it.
3. For `conditionalExpression` and `logicalExpression`, `getLiteralValueAtPath` should add `origin` to `expressionsToBeDeoptimized` at first because `usedBranch` may change from unknown to known.
4. For `memberExpression` , `localVariable` and `callExpressionBase`, the old implementation should be OK

I am working on `hasEffectsOnInteractionAtPath` and will update as soon as I finished.

lukastaegert commented on src/ast/nodes/shared/FunctionBase.ts:
```diff
@@ -27,6 +34,18 @@ import {
 import type { ObjectEntity } from './ObjectEntity';
 import type { PatternNode } from './Pattern';
 
+type FunctionParameterState =
```
I found this TOP/MID/BOTTOM terminology quite confusing. After studying the code, I understand that TOP means "no value yet", MID means "on or several values that are either the same expression or the same known literal value" and BOTTOM means "several different values". Maybe we could improve readability here in one way or another and avoid the arbitrary terminology, maybe even avoid the recreated ad-hoc objects?

First question: Do we really need to initialize the array, or couldn't a value of `undefined` or in general a falsy value carry the same meaning?
Then instead of MID, we just put the expressions themselves into the array, and instead of BOTTOM either put UNKNOWN_EXPRESSION (which is technically also an expression, but of course would have the correct effect on the parameter) or a Symbol value with a helpful name.

liuly0322 replied:
Sorry for lacking the context, I learned these TOP/MID/BOTTOM terminology from constant folding like [here](http://infolab.stanford.edu/~ullman/dragon/w06/lectures/cp.pdf). Sure for readability and performance we should improve the naming and usage.

1. No, we don't need to initialize the array, the default `undefined` is enough
2. Using Symbol for BOTTOM is a good idea 👍since `UNKNOWN_EXPRESSION` is an `ExpressionEntity`, not a node with `parent`, for clarity I would like to create `const UnknownArgument = Symbol('Unknown Argument')

Then: `type FunctionParameterState = ExpressionNode | SpreadElement | typeof UnknownArgument

lukastaegert reviewed (commented):
I just started to review, there will probably some more comments but I am AFK for the next days.

lukastaegert commented:
The fact that it is removing the parameter that is not needed because it statically resolved the condition is really cool: https://rollup-rn22j1qst-rollup-js.vercel.app/repl/?pr=5443&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmZ1bmN0aW9uJTIwZm9vKHgpJTIwJTdCJTVDbiUyMCUyMGlmJTIwKHgpJTIwcmV0dXJuJTIwdHJ1ZSUzQiU1Q24lMjAlMjBlbHNlJTIwcmV0dXJuJTIwZmFsc2UlM0IlNUNuJTdEJTVDbiU1Q25jb25zb2xlLmxvZyhmb28odHJ1ZSkpJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlMkMlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJvdXRwdXQlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiU3RCUyQyUyMnRyZWVzaGFrZSUyMiUzQXRydWUlN0QlN0Q=

liuly0322 commented on test/form/samples/supports-es5-shim/_expected.js:
rollup thinks origin parseInt is not a constructor. [version 4.13.2](https://rollupjs.org/repl/?version=4.13.2&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmNvbnN0JTIwb3JpZ1BhcnNlSW50JTIwJTNEJTIwcGFyc2VJbnQlNUNuJTVDbnBhcnNlSW50JTIwJTNEJTIwZnVuY3Rpb24lMjBwYXJzZUludChzdHIlMkMlMjByYWRpeCklMjAlN0IlNUNuJTVDdGlmJTIwKHRoaXMlMjBpbnN0YW5jZW9mJTIwcGFyc2VJbnQpJTIwJTdCJTIwbmV3JTIwb3JpZ1BhcnNlSW50KCklM0IlMjAlN0QlNUNuJTVDdHJldHVybiUyMG9yaWdQYXJzZUludChzdHIlMkMlMjByYWRpeCklM0IlNUNuJTdEJTNCJTVDbiU1Q25jb25zdCUyMG9yaWdQYXJzZUludDIlMjAlM0QlMjBwYXJzZUludDIlNUNuJTVDbnBhcnNlSW50JTIwJTNEJTIwZnVuY3Rpb24lMjBwYXJzZUludChzdHIlMkMlMjByYWRpeCklMjAlN0IlNUNuJTVDdGlmJTIwKHRoaXMlMjBpbnN0YW5jZW9mJTIwcGFyc2VJbnQpJTIwJTdCJTIwbmV3JTIwb3JpZ1BhcnNlSW50MigpJTNCJTIwJTdEJTVDbiU1Q3RyZXR1cm4lMjBvcmlnUGFyc2VJbnQoc3RyJTJDJTIwcmFkaXgpJTNCJTVDbiU3RCUzQiUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTJDJTIybmFtZSUyMiUzQSUyMm1haW4uanMlMjIlN0QlMkMlN0IlMjJjb2RlJTIyJTNBJTIyJTIyJTJDJTIyaXNFbnRyeSUyMiUzQWZhbHNlJTJDJTIybmFtZSUyMiUzQSUyMm1hdGhzLmpzJTIyJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMm91dHB1dCUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTdEJTJDJTIydHJlZXNoYWtlJTIyJTNBdHJ1ZSU3RCU3RA==) preview. Not sure if this is correct, but has nothing todo with this PR.

liuly0322 commented:
After carefully checking the code several times, I decided to summarize my current optimization conditions.

This PR is closely related to #4510 and #4520. The problem of previous PR is:

> The problem is that for this feature to work correctly, we need to be 100% sure that for every function, we can either track 100% of the calls or if not, deoptimize it. The latter has turned out to become the literal whack-a-mole, causing me to inject so much special logic into the feature that IMO is in no relation to its value. 

In the previous attempt, we tried to trace all calls to `FunctionBase`, which turns out need much special logic to `deOptimize`.

Compared with previous attempts to optimize all `FunctionBase`, this PR once again limits the scope of optimization: we only optimize functions **bound to identifiers** (which may be function expressions or function declarations), and trace All **references to identifiers** to confirm that all uses are function calls. Hopefully, this more conservative strategy will optimize most scenarios while maintaining correctness. (IIFE is a special case and easy to detect, see `FunctionBase.ts`, we don't discuss it now)

(Actually if I don't use this logic, simply optimize all functions will pass the current tests and I cannot find any counterexample, but I still decide to keep it)

This is a two-step process, one is to collect all references to the identifier, and the other is to get the variable corresponding to the identifier from `FunctionBase`.

The first step is actually quite easy. The original `addReference` logic basically completes this. The only difference is that when the module is exported as a whole, the identifier is accessed through the namespace, in `MemberExpression.ts` we should also add the binding.

For the second step, we need to discuss it in two situations. The first is a function expression, and the second is a function declaration. I implemented their logic in `FunctionBase.ts` and `FunctionDeclaration.ts` respectively. That is, either they are bound directly to identifiers, or they are identifiers themselves.

`FunctionBase.ts`:

```typescript
getDeclarationVariable(): Variable | null {
	if (this.parent.type === 'VariableDeclarator') {
		return (this.parent as VariableDeclarator).id.variable ?? null;
	}
	if (this.parent.type === 'ExportDefaultDeclaration') {
		return (this.parent as ExportDefaultDeclaration).variable;
	}
	return null;
}
```

Yes. it does't cover all situations like below:

```typescript
const f = (0, (x) => { return x? x: 0 })
```

But it does track if the function expression that is directly bound to identifiers like below, which should cover most cases:

```typescript
const f1 = () => {}
const f2 = function f() {}

// with name 'foo' special case, that's why we call super.getDeclarationVariable() in `FunctionDeclaration.ts`
export default function foo(x) {
  return x? x: 0
}
```

Otherwise, for `FunctionDeclaration`, we are sure itself is the identifier used by others (I want this to be confirmed, as I only find export default as a special case), so we simply use `id?.variable`

lukastaegert reviewed (commented):
So, sorry it took so long, finally finished my review. This is really great work, I also like the extensive tests you wrote. While I am still not sure we want to keep the "usedPlace" logic forever, I must agree that it makes the feature look rather safe to release.
The alternative would be to make sure that every place where an expression is used makes sure that if there is any possibility that it cannot track all function calls, it properly deoptimizes the function.
However, your approach is rather powerful and covers the 80% case with not too much effort, so maybe we leave it. I added a few small comments but nothing substantial.

liuly0322 commented on src/ast/nodes/shared/CallExpressionBase.ts:
```diff
@@ -59,7 +59,7 @@ export default abstract class CallExpressionBase extends NodeBase implements Deo
 			const { deoptimizableDependentExpressions, expressionsToBeDeoptimized } = this;
 			this.expressionsToBeDeoptimized = EMPTY_SET;
-			this.deoptimizableDependentExpressions = EMPTY_ARRAY as unknown as DeoptimizableEntity[];
+			this.deoptimizableDependentExpressions = [];
```
Sorry I didn't check the one-way logic here, I think it's ok not to track it. reverted.

liuly0322 commented on src/ast/nodes/shared/FunctionBase.ts:
```diff
@@ -57,6 +70,91 @@ export default abstract class FunctionBase extends NodeBase {
+			// only the "this" argument newArguments[0] can be null
+			const argument = newArguments[position + 1] ?? UNDEFINED_EXPRESSION;
+			const parameter = this.params[position];
+			if (!parameter || parameter instanceof RestElement) {
```
Added a comment: `// it's possible that some arguments are empty, so the value is undefined`

liuly0322 commented on src/ast/nodes/shared/FunctionBase.ts:
```diff
@@ -57,6 +70,91 @@ export default abstract class FunctionBase extends NodeBase {
+			});
+			const newValue = argument.getLiteralValueAtPath(EMPTY_PATH, SHARED_RECURSION_TRACKER, {
+				deoptimizeCache() {}
+			});
```
updated

liuly0322 commented on src/ast/nodes/shared/FunctionBase.ts:
```diff
@@ -180,7 +280,27 @@ export default abstract class FunctionBase extends NodeBase {
+	/**
+	 * If the function (expression or declaration) is only used as function calls
+	 */
+	onlyFunctionCallUsed(): boolean {
```
`FuctionDeclaration` rewrites it. changed to protected.

liuly0322 commented on src/ast/nodes/shared/FunctionBase.ts:
```diff
@@ -57,6 +70,91 @@ export default abstract class FunctionBase extends NodeBase {
+	 * updated knownParameters when a call is made to this function
+	 * @param newArguments arguments of the call
+	 */
+	updateKnownArguments(newArguments: InteractionCalledArguments): void {
```
done

liuly0322 commented on src/ast/variables/LocalVariable.ts:
```diff
@@ -93,7 +92,7 @@ export default class LocalVariable extends Variable {
 			this.isReassigned = true;
 			const expressionsToBeDeoptimized = this.expressionsToBeDeoptimized;
-			this.expressionsToBeDeoptimized = EMPTY_ARRAY as unknown as DeoptimizableEntity[];
+			this.expressionsToBeDeoptimized = [];
```
Same like `CallExpressionBase` I did't notice the guard here. done.

liuly0322 commented on src/ast/variables/ExportDefaultVariable.ts:
```diff
@@ -76,7 +92,14 @@ export default class ExportDefaultVariable extends LocalVariable {
+	getOriginalVariable(
+		originalVariableState: OriginalVariableStage = OriginalVariableStage.GENERATE_CHUNK
+	): Variable {
+		if (this.previousOriginalVariableState !== originalVariableState) {
```
https://github.com/rollup/rollup/pull/5443/files#diff-9eb57702aec40721397a4dce9941382b2041e026149026ffe675b301128f2098R11-R12

```typescript
// Between two stages, there might be variable bound change, like `export default a; a = 0`.
// Therefore we clear the cache if the stage changes.
```

We might get a 'wrong'(not complete but sound for function call tracking) original variable in bind reference stage, so just clear the cache if it's in the new stage.

lukastaegert replied:
Interesting. I would have hoped that such scenarios always trigger another tree-shaking pass so that there is always one last pass with the "correct" values.

lukastaegert commented on src/ast/nodes/LogicalExpression.ts:
```diff
@@ -57,21 +57,22 @@ export default class LogicalExpression extends NodeBase implements Deoptimizable
+		}
+		// Request another pass because we need to ensure "include" runs again if
+		// it is rendered
+		context.requestTreeshakingPass();
```
Looks good.

lukastaegert commented:
I just saw you were considering to bind the parameters lazily. This can be done, but basically you have to do it not on "include", but when the identifier becomes part of an "executed code path" (that may or may not be included later, depending on its side effects). To do that, you have to do it in two places: On "hasEffects" (which is a reliable indication that a code path is considered to be executed) AND on "include" (for cases where a code path is directly included without any previous checks). Incidentally, this is exactly when `applyDeoptimizations` is run https://github.com/rollup/rollup/blob/e64f3d8d0cdc561f00d3efe503e3081f81889679/src/ast/nodes/shared/Node.ts#L325-L325
And in a way, adding new call parameters is a potential deoptimization. So you can just extend `applyDeoptimizations` for Identifier and MemberExpression and place the logic there.

liuly0322 commented:
> I just saw you were considering to bind the parameters lazily. This can be done, but basically you have to do it not only on "include", but when the identifier becomes part of an "executed code path" (that may or may not be included later, depending on its side effects). To do that, you have to do it in two places: On "hasEffects" (which is a reliable indication that a code path is considered to be executed) AND on "include" (for cases where a code path is directly included without any previous checks). Incidentally, this is exactly when `applyDeoptimizations` is run
> 
> https://github.com/rollup/rollup/blob/e64f3d8d0cdc561f00d3efe503e3081f81889679/src/ast/nodes/shared/Node.ts#L325-L325
> 
> 
> And in a way, adding new call parameters is a potential deoptimization. So you can just extend `applyDeoptimizations` for Identifier and MemberExpression and place the logic there.

Ah, many thanks, that's exactly what I need! By lazy binding, the `ExportDefaultVariable` is now free of clearing cache. 😉

There is a new small problem in `ParameterVariable`. Ideally, the `isReassigned` change should also do:

```typescript
for (const expression of this.knownExpressionsToBeDeoptimized) {
	expression.deoptimizeCache();
}
```

Otherwise it will block the `expression.deoptimizeCache();` update if we use a early return in `setKnownValue`:

```typescript
if (this.isReassigned) {
	return;
}
```

It's fine to remove it for now, because we only use `isReassigned` to check:

1) assigned to a global variable
2) exported in main

which `isReassigned` change happens before `knownValue` change. Other situations is detected by "usedPlace" mechanism.

However, this should be addressed if we want to remove "usedPlace" check logic in the future, since at that time we will rely on `isReassigned` to trigger cache update.

Do you think we should address it now? I would change `isReassigned` to getter and setter in base class (cannot think of a better solution)

liuly0322 commented:
> > Still one problem to fix, the isReassigned update should also trigger cache update. like this example I would like to make isReassigned a getter or setter, or any other ideas? thanks!
> 
> That is a very good example, if you did not do so already, maybe we can rework this to a function test, I.e. a test that runs itself to see if the correct logic is executed? Maybe change it like [this](https://rollup-hw98h8o4o-rollup-js.vercel.app/repl/?pr=5443&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmZ1bmN0aW9uJTIwZihhKSUyMCU3QiU1Q24lMjAlMjBhc3NlcnQuZXF1YWwoYSUyMCUzRiUyMCdPSyclMjAlM0ElMjAnRkFJTCclMkMlMjAnT0snKSUzQiU1Q24lMjAlMjBhJTIwJTNEJTIwZmFsc2UlM0IlNUNuJTIwJTIwYXNzZXJ0LmVxdWFsKGElMjAlM0YlMjAnRkFJTCclMjAlM0ElMjAnT0snJTJDJTIwJ09LJyklM0IlNUNuJTdEJTVDbiU1Q25mKHRydWUpJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlMkMlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJvdXRwdXQlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiU3RCUyQyUyMnRyZWVzaGFrZSUyMiUzQXRydWUlN0QlN0Q=).
> 
> About changing it to accessors, yes, we can do that. However, this affects all Variables and the field is read much more often than it is written, incurring a performance cost. So why not add a method, e.g. `markReassigned()` that sets the value and which can be overridden. To protect it from unintended writes you can mark it as `readonly` and use a type cast in the setter, e.g.
> 
> ```
> markReassigned() {
>   (this as {isReassigned: boolean}).isReassigned = true;
> }
> ```

As always I learned a lot, really a wonderful experience! The only other place need to be modified is in [deoptimizePath](https://github.com/rollup/rollup/pull/5443/commits/f73a3657e78b5c651454994b2a36e6d7016b3aca#diff-34596877b5032b5f446f8fe47be9f7c4138f1805d46b441d104a24f60754c41bR164-R168) if I understand correctly? The test is also added.

lukastaegert reviewed (approved):
This looks good from my side. Is it ok from your side if I release it?

liuly0322 commented:
> This looks good from my side. Is it ok from your side if I release it?

It looks good to me too🎉. Thanks for your review.

cpojer commented:
I updated [Athena Crisis](https://athenacrisis.com) to the latest version of Vite, and pnpm chose to upgrade Rollup from 4.14.3 to 4.16.1, which comes with this PR. It unfortunately broke my app in production, making it so that `react-router` could no longer match any routes: ![CleanShot_2024-04-22_at_14 06 232x](https://github.com/rollup/rollup/assets/13352/702afff3-2827-44b5-b7b3-c8435abfad19)

Changing `treeshake` to `false` in my Vite config resolves the issue for me, so it is highly likely that this change introduces subtle issues to compiled code. This comment on the `react-router` repo highlights a concrete bug: https://github.com/remix-run/react-router/issues/11480#issuecomment-2068257538

liuly0322 commented:
I'm OK with revert, sorry for the inconvenience.

Using examples from https://github.com/remix-run/react-router/issues/11480#issuecomment-2068204181

It seems that:

```javascript
function flattenRoutes(routes, branches, parentsMeta, parentPath) {
  // It was flattenRoutes(routes, branches=[], ...)
  if (branches === void 0) {
    branches = [];
  }
  ...
}
```

is transformed to:

```javascript
function flattenRoutes(routes, branches, parentsMeta, parentPath) {
  // It was flattenRoutes(routes, branches=[], ...)
  {
    branches = [];
  }
  ...
}
```

I'm trying to figure out the reason, but may not be as fast.

liaochente commented:
> I updated [Athena Crisis](https://athenacrisis.com) to the latest version of Vite, and pnpm chose to upgrade Rollup from 4.14.3 to 4.16.1, which comes with this PR. It unfortunately broke my app in production, making it so that `react-router` could no longer match any routes: ![CleanShot_2024-04-22_at_14 06 232x](https://private-user-images.githubusercontent.com/13352/324339230-702afff3-2827-44b5-b7b3-c8435abfad19.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM3ODAzNTcsIm5iZiI6MTcxMzc4MDA1NywicGF0aCI6Ii8xMzM1Mi8zMjQzMzkyMzAtNzAyYWZmZjMtMjgyNy00NGI1LWI3YjMtYzg0MzVhYmZhZDE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDIyVDEwMDA1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI0ZjFjYzMzNWNlNDJjY2RkNDY3NDliOTZmYWUxYTZiZWM2MzIzZDA0ZTI0MjJmZDkyNDVmNTI1YmJjNGQwOTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.BCyXzLnHjzS4daMeBiAukqw5a7mVIwhtDuBiYKE2I3s)
> 
> Changing `treeshake` to `false` in my Vite config resolves the issue for me, so it is highly likely that this change introduces subtle issues to compiled code. This comment on the `react-router` repo highlights a concrete bug: [remix-run/react-router#11480 (comment)](https://github.com/remix-run/react-router/issues/11480#issuecomment-2068257538)
> 
> Let me know if there is anything I can do to help troubleshoot the issue. It might be best to revert this PR, make a new release, and then rework and fix any issues from this change.

I'm using Vite5, and a temp fix for that issue is:

If you using npm, add to package.json file:

"overrides": {
    "rollup": "4.15.0"
}
If you using pnpm, add to package.json file:

"pnpm": {
    "overrides": {
        "rollup": "4.15.0"
    }
}

brophdawg11 commented:
👋 We just got a bug filed over in react-router for this too - here's some info I found while debugging if it's useful!  It seems like a default param value is being incorrectly transpiled in `flattenRoutes` and end sup being used every time, and overriding the parameter value passed in: https://github.com/remix-run/remix/issues/9271#issuecomment-2069460800

liuly0322 commented:
> 👋 We just got a bug filed over in react-router for this too - here's some info I found while debugging if it's useful! It seems like a default param value is being incorrectly transpiled in `flattenRoutes` and end sup being used every time, and overriding the parameter value passed in: [remix-run/remix#9271 (comment)](https://github.com/remix-run/remix/issues/9271#issuecomment-2069460800)

Thank you for your work! Your conclusion is right, it should be fixed after #5482 merged.

`branches = [];` is a polyfill for default parameter syntax, and I missed this situation: once there is an assignment to `branches`, we shouldn't assume anything about `branches`.

However, if there isn't such a polyfill, 4.16.1 will behave correctly, also chances are there to optimize these default parameters away (though not in current version of rollup, maybe future work).

brophdawg11 commented:
Thank you for the quick turnaround!!

lukastaegert commented:
A revert will be published in the next minutes and we will look at the remaining issues more thoroughly before releasing this feature again.

Shakeskeyboarde commented:
> A revert will be published in the next minutes and we will look at the remaining issues more thoroughly before releasing this feature again.

Love the feature :). Hope you get it sorted and re-released!
